hooks/plan.go

```diff
@@ -0,0 +1,175 @@
+	PlanForAllProcessedBidResponsesStage(endpoint string, account *config.Account) Plan[hookstage.AllProcessedBidResponses]
+	PlanForAuctionResponseStage(endpoint string, account *config.Account) Plan[hookstage.AuctionResponse]
+}
+
+type Plan[T any] []Group[T]
+
+type Group[T any] struct {
+	Timeout time.Duration
+	Hooks   []HookWrapper[T]
+}
+
+type HookWrapper[T any] struct {
+	Module string
+	Code   string
+	Hook   T
+}
+
+func NewExecutionPlanBuilder(hooks config.Hooks, repo HookRepository) ExecutionPlanBuilder {
+	if hooks.Enabled {
+		return PlanBuilder{
+			hooks: hooks,
+			repo:  repo,
+		}
+	}
+	return EmptyPlanBuilder{}
+}
+
+type PlanBuilder struct {
+	hooks config.Hooks
+	repo  HookRepository
+}
+
+func (p PlanBuilder) PlanForEntrypointStage(endpoint string) Plan[hookstage.Entrypoint] {
+	return getMergedPlan(
+		p.hooks,
+		nil,
+		endpoint,
+		StageEntrypoint,
+		p.repo.GetEntrypointHook,
+	)
+}
+
+func (p PlanBuilder) PlanForRawAuctionStage(endpoint string, account *config.Account) Plan[hookstage.RawAuction] {
+	return getMergedPlan(
+		p.hooks,
+		account,
+		endpoint,
+		StageRawAuction,
+		p.repo.GetRawAuctionHook,
+	)
+}
+
+func (p PlanBuilder) PlanForProcessedAuctionStage(endpoint string, account *config.Account) Plan[hookstage.ProcessedAuction] {
+	return getMergedPlan(
+		p.hooks,
+		account,
+		endpoint,
+		StageProcessedAuction,
+		p.repo.GetProcessedAuctionHook,
+	)
+}
+
+func (p PlanBuilder) PlanForBidRequestStage(endpoint string, account *config.Account) Plan[hookstage.BidRequest] {
+	return getMergedPlan(
+		p.hooks,
+		account,
+		endpoint,
+		StageBidRequest,
+		p.repo.GetBidRequestHook,
+	)
+}
+
+func (p PlanBuilder) PlanForRawBidResponseStage(endpoint string, account *config.Account) Plan[hookstage.RawBidResponse] {
+	return getMergedPlan(
+		p.hooks,
+		account,
+		endpoint,
+		StageRawBidResponse,
+		p.repo.GetRawBidResponseHook,
+	)
+}
+
+func (p PlanBuilder) PlanForAllProcessedBidResponsesStage(endpoint string, account *config.Account) Plan[hookstage.AllProcessedBidResponses] {
+	return getMergedPlan(
+		p.hooks,
+		account,
+		endpoint,
+		StageAllProcessedBidResponses,
+		p.repo.GetAllProcessedBidResponsesHook,
+	)
+}
+
+func (p PlanBuilder) PlanForAuctionResponseStage(endpoint string, account *config.Account) Plan[hookstage.AuctionResponse] {
+	return getMergedPlan(
+		p.hooks,
+		account,
+		endpoint,
+		StageAuctionResponse,
+		p.repo.GetAuctionResponseHook,
+	)
+}
+
+type hookFn[T any] func(moduleName string) (T, bool)
+
+func getMergedPlan[T any](
+	cfg config.Hooks,
+	account *config.Account,
+	endpoint, stage string,
+	getHookFn hookFn[T],
+) Plan[T] {
+	accountPlan := cfg.AccountExecutionPlan
+	if account != nil && account.Hooks.ExecutionPlan.Endpoints != nil {
+		accountPlan = account.Hooks.ExecutionPlan
+	}
+
+	plan := getPlan(getHookFn, cfg.HostExecutionPlan, endpoint, stage)
+	plan = append(plan, getPlan(getHookFn, accountPlan, endpoint, stage)...)
```

Comment:
You got it right. The host-level execution plan should be executed regardless of the existence of the account-level execution plan. So what we're doing here is just combining groups of the host-level plan with the groups of the account-level plan.
If the same hook is present in both plans (it looks like it might even present twice in different groups of the same plan) - it would be executed twice or more, but it won't be hidden, the Bid Response returned from the endpoint will contain multiple hook execution results.

In one of the discussions, we decided to leave the whole responsibility of forming the hook execution plan to the host. But I do not see a reason in executing the same hooks multiple times and it's not a problem for us to detect such cases and delete the second hook and/or log some warning despite adding info to the endpoint response. I can add this if you think it would be better.

Reply:
Gotcha. It looks like you're already detecting multiple occurrences of a hook within the host config in `hooks/repo.go#addHook` which is good. Ideally I think it would be helpful to warn if the same module is defined across configs but this could flood the logs. @SyntaxNode @hhhjort perhaps we shouldn't worry about this scenario?